api/storage_service.cc

```diff
@@ -1427,8 +1427,15 @@ void set_snapshot(http_context& ctx, routes& r, sharded<db::snapshot_ctl>& snap_
                     return cm.perform_sstable_scrub(cf.as_table_state(), opts);
                 }, std::make_optional(sstables::compaction_counters{}), reduce_compaction_counters);
             }, std::make_optional(sstables::compaction_counters{}), reduce_compaction_counters);
-        }).then([] (auto f) {
-            if (f->validation_errors) {
+        }).then_wrapped([] (auto f) {
+            if (f.failed()) {
+                auto ex = f.get_exception();
+                if (try_catch<sstables::compaction_aborted_exception>(ex)) {
```

Comment:
Can't we add this to the return value of `perform_task()` too? So it returns a status bit (success, abort, error, etc).

Reply:
I am not sure how status bit could be reduced.

Reply:
Do you mean `perform_task()` should return status bit together with `compaction_stats` or just a status bit?
For the first version it could be something like `std::pair<compaction_stats, compaction_status>` and then mapping it to the `first` member?

Reply:
I thought about something like this:
```
struct compaction_result { // name is debatable
   status status;
   compaction_stats stats;
};
```
We avoid using `std::pair<>` in public APIs, as it is ugly and using a struct is an opportunity to document the purpose with a good name.

Reply:
Currently `perform_task()` throws an exception for aborted operation. Changing it so that it returns `status::aborted` would require changes in behaviour of
```
perform_sstable_scrub_validate_mode()
perform_task_on_all_files()
perform_offstrategy()
compaction_manager::submit()
compaction_manager::run_custom_job()
perform_major_compaction()
```
And all of their callers. Should I really do that? I guess rethrowing or changing of rest api would be needed anyway.

We could also check the retruned value of `perform_task()` in the four latest function and same for `perform_cleanup()` that calls `perform_task_on_all_files()` and if compaction was aborted then rethrow the exception. It doesn't sound reasonable, though.

We could also pass an additional bool parameter set to false on default to `perform_task()` and basing on it throw an exception or just mark it in the status code. That would only obfuscate the code I think.

Or maybe you were thinking about some other solution?

Reply:
Indeed, that is a lot of churn. I still think this would be cleaner, but let's do it as a follow-up.